### PR TITLE
Continued development of scheduler (S^4 and more)

### PR DESCRIPTION
**Pull request by stephtdouglas**
This is an updated version of #137, but @eteq and I decided that I'd mostly take this over for now, so this transfers the responsibility to me. My branch was forked from his, so includes all of his updates from the old PR, plus mine. 

I implemented a scheduler for prioritized observations, which optimizes the best schedule location for each target in order of priority.

### CODE REVIEW

**Comment by eteq** [comment]
I'm going to close #137 now, as this supercedes that.  BUT for anyone who's interested in this topic (esp. @joshwalawender), see that for more discussion on the topic.

Also, @stephtdouglas or @bmorris3 , can you try to summarize some of the discussion from the Python in Astronomy conference about this (or link to a relevant document)?  That way other folks who are involved but weren't at PIA16 can see what's going on.

**Comment by eteq** [comment]
(Also, random tip, @stephtdouglas: if you put ``#137`` in a comment in a description, it automatically links to the issue #137.  So no need to explicitly paste in the full link to an issue/PR)

**Comment by stephtdouglas** [comment]
Our PyAstro16 notes are [here](https://docs.google.com/document/d/1zKo9km8i-lo9TrNDeb6KFkNWWAhXswa2I_41E5RdDd0/edit?usp=sharing). There are a few other notes about things to include in the future, but for now the most crucial thing was trying to nail down our terminology around schedulers. We came up with
- observing block: minimum schedulable block
- priority: number assigned to the observing block by the user (or TAC) that defines its precedence within the set of blocks to be scheduled. Should probably also be on a 0->1 (0=no good, 1=good) scale, or rescaled within the scheduler
- constraint: sets limits and can yield a scores based on inputs. Scores can be boolean or floats (0=no good, 1=good), with a flag in the Constraint call that selects which is used
- score: the value returned by evaluating a constraint
- scorekeeper: assigns a cumulative score to the observing block based on some function that would be applied to all the individual scores (result should be in 0, 1)
- scheduler: the entity that consumes the results of the scorekeeper and the observing blocks and produces a schedule

The schedule that I designed works through targets in order of priority. For now, the scorekeeper is just a loop in each scheduler, but we should consider an external scorekeeper function or class. 

I've posted an example of the PriorityScheduler's results [here](https://github.com/stephtdouglas/astroplan/blob/staging/schedule.ipynb). Note that the current way the schedule is gridded at each loop is leaving a big gap between observations, so that still needs to be fixed. 

I merged the main master branch into my schedule branch to catch up on edits; there were conflicts in constraints.py which I resolved in favor of Erik's edits to generate numerical scores (instead of boolean) when requested.

**Comment by bsipocz** [comment]
@stephtdouglas - Just a technical note: never merge the master branch back to a feature branch but rebase your feature branch on top of the master. It keeps the PRs clear among other things. [0] 
The way to do it is detailed in the astropy dev docs [1], if you're still around I can also help you with it.

[0] https://barro.github.io/2016/02/a-succesful-git-branching-model-considered-harmful/
[1] http://docs.astropy.org/en/latest/development/workflow/development_workflow.html#astropy-guidelines-for-git

**Comment by stephtdouglas** [comment]
OK, I've undone that merge. Someone can help me figure out rebasing later down the line.

File: astroplan/constraints.py [hunk 1]
```diff
@@ -122,22 +164,24 @@ class AltitudeConstraint(Constraint):
...
+        Maximum altitude of the target. `None` indicates no limit.
+    boolean_constraint : bool
+        If True, the constraint is treated as a boolean (True for within the
+        limits and False for outside).  If False, the constraint returns a
+        float on [0, 1], where 0 is the min altitude and 1 is the max.
+    """
+    def __init__(self, min=None, max=None, boolean_constraint=True):
         if min is None:
-            self.min = 0*u.deg
+            self.min = -90*u.deg
```
Comment (bmorris3) [inline astroplan/constraints.py]: What motivates this change? I don't think anyone observes below the horizon, so I don't see a motivation for the minimum altitude to be <0.

File: astroplan/scheduling.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+    priority: integer or float
+        priority of this object in the target list. 1 is highest priority, no maximum
+
+    configuration:
+
+    constraints : sequence of `Constraint`s
+        The constraints to apply to this particular observing block.  Note that
+        constraints applicable to the entire list should go into the Scheduler.
+
```
Comment (bmorris3) [inline astroplan/scheduling.py]: We could improve this docstring a bit.

**Comment by eteq** [comment]
@stephtdouglas - I just remembered right now eteq/astroplan#1 , which I never merged because I never integrated the ideas there and #137.  Do you think you want to include that here?

**Comment by StuartLittlefair** [comment]
@stephtdouglas this is something I (and I guess many others) have an interest in from the point of view of robotic observatories. A well designed scheduling architecture would be a fantastic contribution! One comment on the overall design is that you might like to think about the Minimum Schedulable Block (MSB) being a separate concept/class to the Observing Block. 

The logic is that an observer might need to tie observing blocks (with different targets and constraints) together into one MSB. A simple example might be combining spectroscopic observations with a nearby telluric standard. 

My suggestion is to have Observing Blocks (OBs) defined largely as they are. MSBs could then be created from a list of OBs and Transition Blocks, or an OB could act as a MSB.

**Comment by eteq** [comment]
@StuartLittlefair - just wanted to give you an update on this that's not clear from the PR: there's a Google Summer of Code student (@kvyh) working on finalizing this, so probably there will be another follow-on PR on this to watch out for

**Comment by bmorris3** [comment]
Hey @stephtdouglas, would you mind summarizing what last updates you'd like to make to this PR? I/we can help push it along if you tell us what you wanted to address before we merge it and run with it.

**Comment by stephtdouglas** [comment]
Hey @bmorris3 - the only necessary items left are docstrings and figuring out why the continuous integration checks are failing, and I admittedly haven't had time to figure out how to do the latter. There are still some ideas in this thread from @eteq and @StuartLittlefair that I haven't implemented, but I figured the summer of code student could take a look at those since I didn't have time to start digging into them.

File: astroplan/constraints.py [hunk 3]
```diff
@@ -146,12 +190,14 @@ def __init__(self, min=None, max=None):
             self.max = max
```
Comment (bmorris3) [inline astroplan/constraints.py]: I commented on some other file about this, but meant to comment within the PR here. 

The bug that's failing the ci tests comes from the test that tries to access an attribute on `AltitudeConstraint` that was never set. [Here](https://github.com/astropy/astroplan/blob/0303f571c204e0a867538fcec4f14ceb636c8049/astroplan/constraints.py#L195) you call `self.boolean_constraint`, but the `__init__` method doesn't have `self.boolean_constraint = boolean_constraint` in it (like `AirmassConstraint` does, [for example](https://github.com/astropy/astroplan/blob/0303f571c204e0a867538fcec4f14ceb636c8049/astroplan/constraints.py#L234)).

**Comment by bmorris3** [comment]
The tests now pass, and I'm going to merge this version for continued development by @kvyh.

**Pull request merged**